#### Futuramente essa análise será realizada em pyspark para comparação

In [2]:
import pandas as pd


In [3]:
data = pd.read_csv("stroke_data.csv")

In [3]:
data.head()

0  gender   age  hypertension  heart_disease ever_married      work_type  \
0  1  Female  18.0             0              0           No        Private   
1  2    Male  58.0             1              0          Yes        Private   
2  3  Female  36.0             0              0          Yes       Govt_job   
3  4  Female  62.0             0              0          Yes  Self-employed   
4  5  Female  82.0             0              0          Yes        Private   

  Residence_type  avg_glucose_level    bmi   smoking_status  stroke  
0          Urban              94.19  12.12           smokes       1  
1          Rural             154.24  33.70     never_smoked       0  
2          Urban              72.63  24.70           smokes       0  
3          Rural              85.52  31.20  formerly smoked       0  
4          Rural              59.32  33.20           smokes       1

#### Quantos registros existem no arquivo?

In [5]:
len(data.index)

67135


<kbd>resposta: 67135 </kbd>

In [9]:
data.columns

Index(['0', 'gender', 'age', 'hypertension', 'heart_disease', 'ever_married',
       'work_type', 'Residence_type', 'avg_glucose_level', 'bmi',
       'smoking_status', 'stroke'],
      dtype='object')

In [8]:
#Quantas colunas existem no arquivo? Quantas são numéricas? 
print('DATASET SHAPE: ', data.shape, '(linhas e colunas) \n')
print('-----------')
print('FEATURE DATA TYPES:')
print(data.info())

DATASET SHAPE:  (67135, 12) (linhas e colunas) 

-----------
FEATURE DATA TYPES:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67135 entries, 0 to 67134
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   0                  67135 non-null  int64  
 1   gender             67135 non-null  object 
 2   age                67135 non-null  float64
 3   hypertension       67135 non-null  int64  
 4   heart_disease      67135 non-null  int64  
 5   ever_married       67135 non-null  object 
 6   work_type          67135 non-null  object 
 7   Residence_type     67135 non-null  object 
 8   avg_glucose_level  67135 non-null  float64
 9   bmi                67135 non-null  float64
 10  smoking_status     67135 non-null  object 
 11  stroke             67135 non-null  int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 6.1+ MB
None



<kbd>resposta: 12 e 7 </kbd>

#### Legendas:
- id: unique identifier
-  gender: "Male", "Female" or "Other"
- age: age of the patient
-  hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension
- heart_disease: 0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
- ever_married: "No" or "Yes"
-  work_type: "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
-  Residence_type: "Rural" or "Urban"
-  avg_glucose_level: average glucose level in blood
-  bmi: body mass index
-  smoking_status: "formerly smoked", "never smoked", "smokes" or "Unknown"*
-  stroke: 1 if the patient had a stroke or 0 if not (1 = teve derrame e 0 não teve)
*Note: "Unknown" in smoking_status means that the information is unavailable for this patient


#### No conjunto de dados, quantos pacientes sofreram e não sofreram derrame (stroke), respectivamente?



In [8]:
data['stroke'].value_counts()

1    40287
0    26848
Name: stroke, dtype: int64


<kbd>resposta: 40287 e 26848</kbd>

~~A partir do dataframe, crie uma tabela temporária usando df.createOrReplaceTempView('table') e a seguir use spark.sql para escrever uma consulta SQL que obtenha quantos pacientes tiveram derrame por tipo de trabalho (work_type).~~

#### Quantos pacientes sofreram derrame e trabalhavam respectivamente, no setor privado, de forma independente, no governo e quantas são crianças?

In [19]:
#quantos pacientes tiveram derrame por tipo de trabalho (work_type)
data.filter(items=['work_type','stroke']).groupby(['work_type']).sum()

stroke
work_type            
Govt_job         5164
Never_worked       85
Private         23711
Self-employed   10807
children          520

 
<kbd>resposta: 23711, 10807, 5164, 520. </kbd>
    

~~Escreva uma consulta com spark.sql para determinar a proporção, por gênero~~
#### A maioria dos participantes é:

In [20]:
data['gender'].value_counts()

Female    39530
Male      27594
Other        11
Name: gender, dtype: int64

<kbd>resposta:Female</kbd>

~~Escreva uma consulta com spark.sql para determinar quem tem mais probabilidade de sofrer derrame: hipertensos ou 
    não-hipertensos.Você pode escrever uma consulta para cada grupo. A partir das probabilidades que você obteve~~

#### Quem tem mais probabilidade de sofrer derrame: hipertensos ou não-hipertensos?
lembrando que: hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension

In [250]:
new_data = pd.DataFrame(data, columns = ['hypertension','stroke','age'])
new_data #reduzindo o dataframe para melhor visualizão

hypertension  stroke   age
0                 0       1  18.0
1                 1       0  58.0
2                 0       0  36.0
3                 0       0  62.0
4                 0       1  82.0
...             ...     ...   ...
67130             0       1  18.0
67131             0       0  29.0
67132             0       1  24.0
67133             0       1  77.0
67134             0       1  76.0

[67135 rows x 3 columns]

In [30]:
import seaborn as sns 
import matplotlib.pyplot as plt 

In [21]:
data.loc[data['hypertension']==1,['stroke']].value_counts()


stroke
1         8817
0         2200
dtype: int64

<kbd>ou seja, 8817 pacientes com hipertensao tiveram derrame (stroke)o que é (80%). 2200 pacientes não tiveram. (20%)
</kbd>


##### **(obs: Fazer analise  utilizando as métricas de estatistica depois)**

In [30]:
data.loc[data['hypertension']==0,['stroke']]

stroke
0           1
2           0
3           0
4           1
5           0
...       ...
67130       1
67131       0
67132       1
67133       1
67134       1

[56118 rows x 1 columns]

In [33]:

(data.loc[data['hypertension']==0,['stroke']]).value_counts()

stroke
1         31470
0         24648
dtype: int64

In [35]:

(data.loc[data['hypertension']==0,['stroke']]).count()

stroke    56118
dtype: int64

<kbd>ou seja, 24648 pacientes sem hipertensao não tem stroke (43.9%) e 31470 tem (56 %)
Neste caso, mesmo sem hipertensao o número de casos de stroke é maior.</kbd>


~~Escreva uma consulta com spark.sql que determine o número de pessoas que sofreram derrame por idade~~
#### Com qual idade o maior número de pessoas do conjunto de dados sofreu derrame?

In [42]:
data.loc[data['stroke']==1,['age']].value_counts()

age  
79.00    2916
78.00    2279
80.00    1858
81.00    1738
82.00    1427
         ... 
13.00     131
12.00     119
11.00      95
10.00      62
1.32        1
Length: 74, dtype: int64

<kbd>resposta:79.00 </kbd>

##### ~~Usando a API de dataframes~~ 
#### Determine quantas pessoas sofreram derrames após os 50 anos.

In [40]:
particao =data.loc[data['stroke']==1,['age']]
len(particao.loc[particao['age']>50])

28938


<kbd>resposta:28938 </kbd>

~~Usando spark.sql~~
#### Determine qual o nível médio de glicose para pessoas que, respectivamente, sofreram e não sofreram derrame.

In [53]:
data.loc[data['stroke']==1,['avg_glucose_level']].mean()

avg_glucose_level    119.95307
dtype: float64

In [54]:
data.loc[data['stroke']==0,['avg_glucose_level']].mean()

avg_glucose_level    103.602731
dtype: float64

<kbd>resposta:119 e 103 </kbd>

#### Qual é o BMI (IMC = índice de massa corpórea) médio de quem sofreu e não sofreu derrame?

In [57]:
data.loc[data['stroke']==1,['bmi']].mean()

bmi    29.942491
dtype: float64

In [58]:
data.loc[data['stroke']==0,['bmi']].mean()

bmi    27.989679
dtype: float64

<kbd>resposta:29,94 e 27,99 </kbd>

### Parte 2 - Machine learning

#### 1- Crie um modelo de árvore de decisão que prevê a chance de derrame (stroke) a partir das variáveis contínuas/categóricas: idade, BMI, hipertensão, doença do coração, nível médio de glicose.
#### 2- Qual a acurácia de um modelo construído?


In [4]:
new_df = data[['gender','age','hypertension','heart_disease','avg_glucose_level','bmi','smoking_status','stroke']]

In [5]:
new_df.loc[new_df.stroke==1,'stroke']='Teve Derrame'
new_df.loc[new_df.stroke==0,'stroke']='Não Teve Derrame'
new_df.head()

C:\Users\la-re\AppData\Local\Temp\ipykernel_19920\262867914.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df.loc[new_df.stroke==1,'stroke']='Teve Derrame'


gender   age  hypertension  heart_disease  avg_glucose_level    bmi  \
0  Female  18.0             0              0              94.19  12.12   
1    Male  58.0             1              0             154.24  33.70   
2  Female  36.0             0              0              72.63  24.70   
3  Female  62.0             0              0              85.52  31.20   
4  Female  82.0             0              0              59.32  33.20   

    smoking_status            stroke  
0           smokes      Teve Derrame  
1     never_smoked  Não Teve Derrame  
2           smokes  Não Teve Derrame  
3  formerly smoked  Não Teve Derrame  
4           smokes      Teve Derrame

In [431]:
X = new_df.iloc[:,1:6].values # previsores
y = new_df.iloc[:,7].values #alvo

In [432]:
X

array([[ 18.  ,   0.  ,   0.  ,  94.19,  12.12],
       [ 58.  ,   1.  ,   0.  , 154.24,  33.7 ],
       [ 36.  ,   0.  ,   0.  ,  72.63,  24.7 ],
       ...,
       [ 24.  ,   0.  ,   0.  , 141.18,  23.3 ],
       [ 77.  ,   0.  ,   0.  , 199.71,  36.2 ],
       [ 76.  ,   0.  ,   1.  , 202.96,  31.8 ]])

In [376]:
y

array(['Teve Derrame', 'Não Teve Derrame', 'Não Teve Derrame', ...,
       'Teve Derrame', 'Teve Derrame', 'Teve Derrame'], dtype=object)

In [450]:
#dividindo o dataset em treino e teste
from sklearn.model_selection import train_test_split
X_treinamento, X_teste, y_treinamento, y_test = train_test_split(X, y, test_size = 0.25, random_state=0)

In [451]:
#ajustando a escala dos dados
from sklearn.preprocessing import StandardScaler
scaler_x = StandardScaler()
X = scaler_x.fit_transform(X)

In [452]:
X_treinamento.shape

(50351, 5)

In [453]:
from sklearn.tree import DecisionTreeClassifier

In [454]:
arvore_stroke = DecisionTreeClassifier(criterion ='entropy', max_depth =20) #pesquisar os critérios
arvore_stroke.fit(X_treinamento, y_treinamento)


DecisionTreeClassifier(criterion='entropy', max_depth=20)

In [455]:
new_df.columns

Index(['gender', 'age', 'hypertension', 'heart_disease', 'avg_glucose_level',
       'bmi', 'smoking_status', 'stroke'],
      dtype='object')

In [456]:
arvore_stroke.feature_importances_  # Grau de importancia de cada variável

array([0.38735572, 0.01173274, 0.01834067, 0.34800074, 0.23457013])

**A feature mais importante foi a idade seguida da avg_glucose_level**

In [457]:
arvore_stroke.classes_

array(['Não Teve Derrame', 'Teve Derrame'], dtype=object)

In [458]:
previsoes = arvore_stroke.predict(X_teste)

In [459]:
previsoes

array(['Não Teve Derrame', 'Teve Derrame', 'Teve Derrame', ...,
       'Teve Derrame', 'Não Teve Derrame', 'Não Teve Derrame'],
      dtype=object)

In [460]:
y_test

array(['Teve Derrame', 'Teve Derrame', 'Teve Derrame', ...,
       'Teve Derrame', 'Não Teve Derrame', 'Teve Derrame'], dtype=object)

In [461]:
from sklearn.metrics import accuracy_score
taxa_acerto = accuracy_score(y_test, previsoes)
taxa_acerto

0.7300405147759771

<kbd>resposta:Menor que 75% </kbd>

In [141]:
import matplotlib.pyplot as plt
from sklearn import tree

In [ ]:
# #plot da arvore
# figura, eixos = plt.subplots(ncols=1,figsize=(11,11))
# tree.plot_tree(arvore_stroke, feature_names= ['age', 'hypertension', 'heart_disease', 'avg_glucose_level', 'bmi'])

#### Adicione ao modelo as variáveis categóricas: genero e status de fumante.

In [398]:
#fazendo one hot enconding. 
new_df2 = pd.get_dummies(new_df.smoking_status, prefix='Smoke')

new_df2.head()


Smoke_formerly smoked  Smoke_never_smoked  Smoke_smokes
0                      0                   0             1
1                      0                   1             0
2                      0                   0             1
3                      1                   0             0
4                      0                   0             1

In [399]:
new_df3 = pd.get_dummies(new_df.gender, prefix='Gender')
new_df3.head()

Gender_Female  Gender_Male  Gender_Other
0              1            0             0
1              0            1             0
2              1            0             0
3              1            0             0
4              1            0             0

In [400]:
#juntando dataframes
new_df4= new_df.join(new_df2)
new_df5= new_df4.join(new_df3)

In [401]:
new_df5= new_df5.drop(columns=['smoking_status', 'gender'])
new_df5.head()

age  hypertension  heart_disease  avg_glucose_level    bmi  \
0  18.0             0              0              94.19  12.12   
1  58.0             1              0             154.24  33.70   
2  36.0             0              0              72.63  24.70   
3  62.0             0              0              85.52  31.20   
4  82.0             0              0              59.32  33.20   

             stroke  Smoke_formerly smoked  Smoke_never_smoked  Smoke_smokes  \
0      Teve Derrame                      0                   0             1   
1  Não Teve Derrame                      0                   1             0   
2  Não Teve Derrame                      0                   0             1   
3  Não Teve Derrame                      1                   0             0   
4      Teve Derrame                      0                   0             1   

   Gender_Female  Gender_Male  Gender_Other  
0              1            0             0  
1              0            1             0  
2              1            0             0  
3              1            0             0  
4              1            0             0

In [402]:
#ordenando as colunas para facilitar
new_df5 = new_df5[['age','hypertension','heart_disease','avg_glucose_level','bmi','Gender_Female','Gender_Male','Gender_Other','Smoke_formerly smoked','Smoke_never_smoked','Smoke_smokes','stroke']]

In [403]:
new_df5

age  hypertension  heart_disease  avg_glucose_level    bmi  \
0      18.0             0              0              94.19  12.12   
1      58.0             1              0             154.24  33.70   
2      36.0             0              0              72.63  24.70   
3      62.0             0              0              85.52  31.20   
4      82.0             0              0              59.32  33.20   
...     ...           ...            ...                ...    ...   
67130  18.0             0              0             109.06  22.10   
67131  29.0             0              0              73.02  32.10   
67132  24.0             0              0             141.18  23.30   
67133  77.0             0              0             199.71  36.20   
67134  76.0             0              1             202.96  31.80   

       Gender_Female  Gender_Male  Gender_Other  Smoke_formerly smoked  \
0                  1            0             0                      0   
1                  0            1             0                      0   
2                  1            0             0                      0   
3                  1            0             0                      1   
4                  1            0             0                      0   
...              ...          ...           ...                    ...   
67130              0            1             0                      0   
67131              1            0             0                      0   
67132              1            0             0                      0   
67133              1            0             0                      1   
67134              0            1             0                      1   

       Smoke_never_smoked  Smoke_smokes            stroke  
0                       0             1      Teve Derrame  
1                       1             0  Não Teve Derrame  
2                       0             1  Não Teve Derrame  
3                       0             0  Não Teve Derrame  
4                       0             1      Teve Derrame  
...                   ...           ...               ...  
67130                   0             1      Teve Derrame  
67131                   1             0  Não Teve Derrame  
67132                   0             1      Teve Derrame  
67133                   0             0      Teve Derrame  
67134                   0             0      Teve Derrame  

[67135 rows x 12 columns]

In [418]:
new_df5.columns

Index(['age', 'hypertension', 'heart_disease', 'avg_glucose_level', 'bmi',
       'Gender_Female', 'Gender_Male', 'Gender_Other', 'Smoke_formerly smoked',
       'Smoke_never_smoked', 'Smoke_smokes', 'stroke'],
      dtype='object')

In [419]:
X2 = new_df5.iloc[:,0:11].values # previsores
y2 = new_df5.iloc[:,11].values #alvo

In [420]:
X2

array([[18.,  0.,  0., ...,  0.,  0.,  1.],
       [58.,  1.,  0., ...,  0.,  1.,  0.],
       [36.,  0.,  0., ...,  0.,  0.,  1.],
       ...,
       [24.,  0.,  0., ...,  0.,  0.,  1.],
       [77.,  0.,  0., ...,  1.,  0.,  0.],
       [76.,  0.,  1., ...,  1.,  0.,  0.]])

In [408]:
y2

array(['Teve Derrame', 'Não Teve Derrame', 'Não Teve Derrame', ...,
       'Teve Derrame', 'Teve Derrame', 'Teve Derrame'], dtype=object)

In [409]:
print(new_df5.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67135 entries, 0 to 67134
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   age                    67135 non-null  float64
 1   hypertension           67135 non-null  int64  
 2   heart_disease          67135 non-null  int64  
 3   avg_glucose_level      67135 non-null  float64
 4   bmi                    67135 non-null  float64
 5   Gender_Female          67135 non-null  uint8  
 6   Gender_Male            67135 non-null  uint8  
 7   Gender_Other           67135 non-null  uint8  
 8   Smoke_formerly smoked  67135 non-null  uint8  
 9   Smoke_never_smoked     67135 non-null  uint8  
 10  Smoke_smokes           67135 non-null  uint8  
 11  stroke                 67135 non-null  object 
dtypes: float64(3), int64(2), object(1), uint8(6)
memory usage: 3.5+ MB
None


In [421]:

X2_treinamento, X2_teste, y2_treinamento, y2_test = train_test_split(X2, y2, test_size = 0.25, random_state=0)
scaler_x2 = StandardScaler()
X2 = scaler_x2.fit_transform(X2)
X2_treinamento.shape


(50351, 11)

In [422]:
arvore_stroke2 = DecisionTreeClassifier(criterion ='entropy', max_depth =20) #olhar isso
arvore_stroke2.fit(X2_treinamento, y2_treinamento)
previsoes2 = arvore_stroke2.predict(X2_teste)


In [412]:
new_df5.columns

Index(['age', 'hypertension', 'heart_disease', 'avg_glucose_level', 'bmi',
       'Gender_Female', 'Gender_Male', 'Gender_Other', 'Smoke_formerly smoked',
       'Smoke_never_smoked', 'Smoke_smokes', 'stroke'],
      dtype='object')

In [423]:
arvore_stroke2.feature_importances_  # Grau de importancia de cada variável


array([0.16109844, 0.01164923, 0.01452833, 0.20570726, 0.15729982,
       0.00844469, 0.00559637, 0.        , 0.01772869, 0.37437751,
       0.04356964])

**A feature mais importante é Smoke**

In [424]:
arvore_stroke2.classes_

array(['Não Teve Derrame', 'Teve Derrame'], dtype=object)

In [425]:
previsoes2

array(['Teve Derrame', 'Teve Derrame', 'Teve Derrame', ...,
       'Teve Derrame', 'Não Teve Derrame', 'Teve Derrame'], dtype=object)

In [426]:
y2_test

array(['Teve Derrame', 'Teve Derrame', 'Teve Derrame', ...,
       'Teve Derrame', 'Não Teve Derrame', 'Teve Derrame'], dtype=object)

#### A acurácia (qualidade) do modelo aumentou para

In [428]:

taxa_acerto2 = accuracy_score(y2_test, previsoes2)
taxa_acerto2

0.8731530028598665

<kbd>resposta:Acima de 85% </kbd>